File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/video/OutputOptionsImpl.kt
package dev.hebei.camerax_android.video

import android.location.Location
import androidx.camera.video.OutputOptions
import dev.hebei.camerax_android.CameraXImpl
import dev.hebei.camerax_android.PigeonApiOutputOptionsApi

class OutputOptionsImpl(impl: CameraXImpl): PigeonApiOutputOptionsApi(impl) {
    override fun getDurationLimitMillis(pigeon_instance: OutputOptions): Long {
        return pigeon_instance.durationLimitMillis
    }

    override fun getFileSizeLimit(pigeon_instance: OutputOptions): Long {
        return pigeon_instance.fileSizeLimit
    }

    override fun getLocation(pigeon_instance: OutputOptions): Location? {
        return pigeon_instance.location
    }
}

File: camerax_android/lib/src/video/output_options_impl.dart
import 'package:camerax_android/src/api.dart';
import 'package:camerax_android/src/common.dart';
import 'package:camerax_android/src/video.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

abstract base class OutputOptionsImpl implements OutputOptions {
  OutputOptionsProxyApi get api;

  @override
  Future<Duration> getDurationLimit() =>
      api.getDurationLimitMillis().then((e) => Duration(milliseconds: e));

  @override
  Future<int> getFileSizeLimit() => api.getFileSizeLimit();

  @override
  Future<Location?> getLocation() => api.getLocation().then((e) => e?.impl);
}

extension OutputOptionsProxyApiX on OutputOptionsProxyApi {
  OutputOptions get impl {
    final api = this;
    if (api is FileOutputOptionsProxyApi) {
      return FileOutputOptionsImpl.internal(api);
    } else {
      throw TypeError();
    }
  }
}
